### PR TITLE
Add BVH creation

### PR DESCRIPTION
**Pull request by bradgrantham**
Notably adds Branch and Leaf types, bvh.[ch] adding the BVH tree creator, and range.[ch] providing range functionality.

Set NO_BVH environment variable to 1 (or anything) to not create or use a BVH.

### CODE REVIEW

File: src/branch.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        child1 = &branch->fartherChild;
+    } else {
+        child1 = &branch->closerChild;
+        child0 = &branch->fartherChild;
+    }
+
+    int hitsBox;
+    int result = 0;
+
+    // XXX hitsBox = child0 box intersect
```
Comment (jaggies) [inline src/branch.c]: What is XXX? TODO?

File: src/branch.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        result += child1->shape->op->intersect(child1->shape, ray, hit);
+
+    return result;
+}
+
+static
+void Branch_normal(struct Shape* shape, Hit* hit, Vec3 *n) {
+    Branch* branch = (Branch*) shape;
+
+    // ???
```
Comment (jaggies) [inline src/branch.c]: Please remove these.

File: src/hit.h [hunk 3]
```diff
@@ -10,7 +10,10 @@
 
 #include "vec2.h"
 #include "vec3.h"
-#include "shape.h"
+#include "ray.h"
+// #include "shape.h"
```
Comment (jaggies) [inline src/hit.h]: Please remove this

File: src/branch.c [hunk 4]
```diff
@@ -11,12 +11,56 @@
 #include "shader.h"
 #include "hit.h"
 
+void range_init(Vec2* r) {
+    r->x = -100000000.0;
```
Comment (jaggies) [inline src/branch.c]: Please declare constants (+-MAX_RANGE?)

File: src/branch.c [hunk 5]
```diff
@@ -11,12 +11,56 @@
...
+extern long intersections;
+
 static
 int Branch_intersect(struct Shape* shape, Ray* ray, Hit* hit) {
 
     if(shape == hit->ignore)
         return 0;
 
+    intersections++;
+
```
Comment (jaggies) [inline src/branch.c]: #ifdef PROFILE

File: src/leaf.c [hunk 6]
```diff
@@ -14,12 +14,17 @@
...
+// XXX
+extern long intersections;
+
 static
 int Leaf_intersect(struct Shape* shape, Ray* ray, Hit* hit) {
 
     if(shape == hit->ignore)
         return 0;
 
+    intersections++;
```
Comment (jaggies) [inline src/leaf.c]: #ifdef PROFILE

File: src/sphere.c [hunk 7]
```diff
@@ -13,12 +13,17 @@
...
+// XXX
+extern long intersections;
+
 static
 int intersect(struct Shape* shape, Ray* ray, Hit* hit) {
 
     if(shape == hit->ignore)
         return 0;
 
+    intersections++;
```
Comment (jaggies) [inline src/sphere.c]: #ifdef PROFILE

File: src/triangle.c [hunk 8]
```diff
@@ -18,12 +18,17 @@ struct TriangleHit {
...
+// XXX
+extern long intersections;
+
 static
 int intersect(struct Shape* shape, Ray* ray, Hit *hit) {
 
     if(shape == hit->ignore)
         return 0;
 
+    intersections++;
```
Comment (jaggies) [inline src/triangle.c]: #ifdef PROFILE

File: src/triangle.h [hunk 9]
```diff
@@ -13,7 +13,7 @@
 typedef struct Triangle {
     ShapeOps* op;
     Shader* shader;
-    Vec3  point[1];
+    Vec3  point[3];
```
Comment (jaggies) [inline src/triangle.h]: #1 This should be changed back and use point[0] + edge[n] to get the original points.

**Comment by bradgrantham** [comment]
Fixes #2

**Pull request merged**